taxjar/taxjar-magento2-extension
# INT-1974: Backup Rates Async refactor

### Context

Some users where nexus includes many U.S. states have ran into an issue where backup rates cannot be stored/sync'd due to process timeout with the current backup rates feature.

### Description

This refactors our existing "Backup rates" feature to utilize a message queue system processing chunks of backup rate data instead of relying on one synchronous process to handle creating/updating/deleting all backup rates. Includes Unit/Integration test coverage.

### Performance

These changes provide a net performance increase for the application since backup rates are processed in background.

### Testing

To test rate creation:
1. Activate backup rates feature.
2. Observe notification that backup rates jobs have been queued (either at top of page or in bulk action log)
3. Depending on how M2 is configured to process message queue, either wait for cron, or manually start consumer processes by entering `bin/magento queue:consumers:start taxjar.backup_rates.create` and `bin/magento queue:consumers:start taxjar.backup_rates.delete`.
4. Once consumers process messages and actions are complete, observe that number of rates loaded matches number of rates expected.

To test rate deletion:
1. With backup rates feature enabled, disable by selecting "No" and saving configuration.
2. Observe notification that backup rates have been queued for removal.
3. Depending on how M2 is configured to process message queue, either wait for cron, or manually start consumer processes by entering `bin/magento queue:consumers:start taxjar.backup_rates.create` and `bin/magento queue:consumers:start taxjar.backup_rates.delete`.
4. Once consumers process messages and actions are complete, observe that all previously created rates and calculations have been removed from the database.

#### Versions
